File: src/playground/notification/skype.clj
(ns playground.notification.skype
  (:require [taoensso.timbre :as timbre]
            [clj-http.client :as http]
            [cheshire.core :as json]
            [playground.data.config :as c]
            [playground.utils.utils :as utils]
            [clojure.string :as string])
  (:import (org.apache.commons.lang3 StringEscapeUtils)))


;; =====================================================================================================================
;; Settings and util functions
;; =====================================================================================================================
(defn- config [notifier] (-> notifier :config :skype))


(defn get-access-token [id key]
  (let [url "https://login.microsoftonline.com/common/oauth2/v2.0/token"
        data {"client_id"     id
              "scope"         "https://api.botframework.com/.default"
              "grant_type"    "client_credentials"
              "client_secret" key}
        resp (http/post url {:form-params data})
        body (json/parse-string (:body resp) true)
        access-token (:access_token body)]
    access-token))


(defn send-msg [chat-id access-token message]
  (let [url (str "https://apis.skype.com/v2/conversations/" chat-id "/activities")
        data {:message {:content message}}
        headers {"Authorization" (str "Bearer " access-token)}
        resp (http/post url {:body    (json/generate-string data)
                             :headers headers})]
    resp))


(defn send-message [{:keys [id key chat-id]} message]
  (try
    (let [access-token (get-access-token id key)]
      (send-msg chat-id access-token message))
    (catch Exception e
      (timbre/error "Skype send message error: " message e (.getStackTrace e)))))


(defn send-release-message [conf version message]
  (when (and (:release-chat-id conf)
             (utils/released-version? version)
             (= (c/prefix) "prod"))
    (send-message (assoc conf :chat-id (:release-chat-id conf)) message)))


(defn font [text & [color size]]
  (str "<font "
       (when color (str "color=\"" color "\" "))
       (when size (str "size=\"" size "px\"")) ">"
       text "</font>"))


(defn b [text] (str "<b>" text "</b>"))
(defn u [text] (str "<u>" text "</u>"))
(defn i [text] (str "<i>" text "</i>"))
(defn code [text] (str "{code}\n" text "\n{code}"))


;; =====================================================================================================================
;; Notifications functions
;; =====================================================================================================================
(defn application-start [notifier]
  (let [msg (str "[PG " (c/prefix) "] " (-> "application has been started" (font "#4183C4") b) "\n")]
    (send-message (config notifier) msg)))


(defn start-version-building [notifier project {author         :author
                                                commit-message :message
                                                version        :name
                                                commit         :commit} queue-index]
  (let [msg (str "[PG " (c/prefix) "] #" queue-index " " (b (str project "/" version))
                 " \"" (StringEscapeUtils/escapeHtml4 commit-message) "\" @" author " (" (subs commit 0 7) ") - "
                 (-> "start" (font "#4183C4")) "\n")]
    (send-message (config notifier) msg)
    (send-release-message (config notifier) version msg)))


(defn complete-version-building [notifier project {author         :author
                                                   commit-message :message
                                                   version        :name
                                                   commit         :commit} queue-index]
  (let [msg (str "[PG " (c/prefix) "] #" queue-index " " (b (str project "/" version))
                 " \"" (StringEscapeUtils/escapeHtml4 commit-message) "\" @" author " (" (subs commit 0 7) ") - "
                 (-> "complete" (font "#36a64f")) "\n")]
    (send-message (config notifier) msg)
    (send-release-message (config notifier) version msg)))


(defn complete-version-building-error [notifier project {author         :author
                                                         commit-message :message
                                                         version        :name
                                                         commit         :commit} queue-index e]
  (let [msg (str "[PG " (c/prefix) "] #" queue-index " " (b (str project "/" version))
                 " \"" (StringEscapeUtils/escapeHtml4 commit-message) "\" @" author " (" (subs commit 0 7) ") - "
                 (-> "failed" (font "#d00000") b) "\n"
                 (when e
                   (-> (utils/format-exception e) (font "#777777" 11) i)))]
    (send-message (config notifier) msg)
    (send-release-message (config notifier) version msg)))


;(defn complete-building-with-errors [notifier branches queue-index e]
;  (let [msg (str "#" queue-index " api " (-> (c/prefix) (font "#cc0066" 11) u) " - " (-> "error during processing!" (font "#d00000") b) "\n"
;                 (when (seq branches)
;                   (str (b "Branches: ") (string/join ", " branches)))
;                 (when e
;                   (-> (utils/format-exception e) (font "#777777" 11) i)))]
;    (send-message (config notifier) msg)))

File: .travis/skype.py
#!/usr/bin/env python
# coding=utf-8
import sys
import os
import requests
import subprocess


red_color = "#7c0000"
green_color = "#007c00"


def __send_notification(skype_id, skype_key, chat_id, msg):
    def __get_access_token(id, secret):
        url = 'https://login.microsoftonline.com/common/oauth2/v2.0/token'
        data = {'client_id': id,
                'scope': 'https://api.botframework.com/.default',
                'grant_type': 'client_credentials',
                'client_secret': secret}
        try:
            r = requests.post(url, data=data)
            if r.status_code == requests.codes.ok:
                return r.json()['access_token']
        except Exception as e:
            print e
    def __send_msg(message, chat_id, token):
        url = ('https://apis.skype.com/v2/conversations/{}/activities'.format(chat_id))
        data = {'message': {'content': message}}
        headers = {'Authorization': 'Bearer {}'.format(token)}
        if token:
            try:
                return requests.post(url, json=data, headers=headers)
            except Exception as e:
                print e

    token = __get_access_token(skype_id, skype_key)
    __send_msg(msg, chat_id, token)


def commit_hash():
  return subprocess.check_output(["git", "rev-parse", "--short", "HEAD"]).strip()


def commit_msg():
  return subprocess.check_output(["git", "log", "-1", "--pretty=%B"]).strip()


def commit_author():
  return subprocess.check_output(["git", "log", "-1", "--pretty=%an"]).strip()


def colorize(color, s):
  return "<b><font color='" + color + "'>" + s + "</font></b>"


def failed_msg():
  return colorize(red_color, "failed")


def success_msg():
  return colorize(green_color, "complete")


def notify(skype_id, skype_key, skype_chat_id, company_repo, branch, build_id, build_num, success):
  msg = '[PG travis] #{build_num} <b>{branch}</b> "{commit_msg}" @{commit_author} ({commit_hash}) - {msg}'\
         .format(build_num=build_num, branch=branch, commit_msg=commit_msg(), commit_author=commit_author(), 
                 commit_hash=commit_hash(), msg=(success_msg() if int(success) else failed_msg()))
  report = "<a href=\"https://travis-ci.org/" + company_repo + "/builds/" + str(build_id) + "\">See report</a>"
  __send_notification(skype_id, skype_key, skype_chat_id, msg + "\n" + report)


notify(sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4], sys.argv[5], sys.argv[6] , sys.argv[7], sys.argv[8])
